lib/using_restricted_array.rb

```diff
@@ -6,56 +6,152 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def length(array)
-  raise NotImplementedError
-end
+  i = 0
+
+  until array[i] == nil
+    i += 1
+  end
+
+    return i
+  end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def print_array(array)
-  raise NotImplementedError
+  i = 0
+
+  until array[i] == nil
+    puts i
+    i += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity:
+#   sorted: O(n)
+#   unsorted: O(n)
+# Space complexity: O(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  i = 0
+
+  while i < length
+    if array[i] == value_to_find
+      return true
+    else
+      i += 1
+    end
+  end
+  return false
 end
 
+
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity:
+#   sorted: O(1)
+#   unsorted: O(n)
+# Space complexity: O(1)
 def find_largest(array, length)
-  raise NotImplementedError
-end
+  i = 0
+
+  if length  == 0
+    return nil
+  else
+    max = array[0]
+
+    while i < length
+      if array[i] > max
+        max = array[i]
+      end
+      i += 1
+    end
+
+  return max
+  end
+  end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity:
+#   sorted: O(1)
+#   unsorted: O(n)
+# Space complexity: O(1)
 def find_smallest(array, length)
-  raise NotImplementedError
+  i = 0
+
+  if length  == 0
+    return nil
+  else
+    min = array[0]
+    while i < length
+      if array[i] < min
+        min = array[i]
+      end
+      i += 1
+    end
+
+    return min
+  end
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(1)
 def reverse(array, length)
```

Comment:
👍 , this works but since you're making an auxiliary array it has O(n) space complexity.  

Can you think of a way to do this without the other array?